c2rust-refactor/src/scripting/to_lua_ast_node.rs

```diff
@@ -0,0 +1,362 @@
+        self.0.borrow_mut().walk(visitor);
+    }
+}
+
+unsafe impl Send for LuaAstNode<P<Item>> {}
+#[allow(unused_doc_comments)]
+impl UserData for LuaAstNode<P<Item>> {
+    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
+        methods.add_method("get_kind", |_lua_ctx, this, ()| {
+            Ok(this.0.borrow().node.ast_name())
+        });
+
+        methods.add_method("get_id", |lua_ctx, this, ()| {
+            Ok(this.0.borrow().id.to_lua(lua_ctx))
+        });
+
+        methods.add_method("get_ident", |_lua_ctx, this, ()| {
+            Ok(this.0.borrow().ident.to_string())
+        });
+
+        methods.add_method("set_ident", |_lua_ctx, this, ident: String| {
+            this.0.borrow_mut().ident = Ident::from_str(&ident);
+            Ok(())
+        });
+
+        /// Visit statements
+        // @function visit_stmts
+        // @tparam function(LuaAstNode) callback Function to call when visiting each statement
+        methods.add_method("visit_stmts", |lua_ctx, this, callback: Function| {
+            visit_nodes(&**this.borrow(), |node: &Stmt| {
+                callback.call::<_, ()>(node.clone().to_lua(lua_ctx))
+                .unwrap_or_else(|e| panic!("Lua callback failed in visit_stmts: {}", DisplayLuaError(e)));
+            });
+            Ok(())
+        });
+
+        methods.add_method("visit_items", |lua_ctx, this, callback: Function| {
+            visit_nodes(&**this.borrow(), |node: &Item| {
+                callback.call::<_, ()>(P(node.clone()).to_lua(lua_ctx))
+                .unwrap_or_else(|e| panic!("Lua callback failed in visit_items: {}", DisplayLuaError(e)));
+            });
+            Ok(())
+        });
+
+        methods.add_method("visit_foreign_items", |lua_ctx, this, callback: Function| {
+            visit_nodes(&**this.borrow(), |node: &ForeignItem| {
+                callback.call::<_, ()>(P(node.clone()).to_lua(lua_ctx))
+                .unwrap_or_else(|e| panic!("Lua callback failed in visit_foreign_items: {}", DisplayLuaError(e)));
+            });
+            Ok(())
+        });
+
+        methods.add_method("get_node", |lua_ctx, this, ()| {
+            match this.0.borrow().node.clone() {
+                ItemKind::Use(e) => Ok(e.to_lua(lua_ctx)),
+                node => Err(Error::external(format!("Item node {:?} not implemented yet", node))),
+            }
+        });
+    }
+}
+
+unsafe impl Send for LuaAstNode<P<ForeignItem>> {}
+impl UserData for LuaAstNode<P<ForeignItem>> {
+    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
+        methods.add_method("get_kind", |_lua_ctx, this, ()| {
+            Ok(this.0.borrow().node.ast_name())
+        });
+
+        methods.add_method("get_id", |lua_ctx, this, ()| {
+            Ok(this.0.borrow().id.to_lua(lua_ctx))
+        });
+
+        methods.add_method("get_ident", |_lua_ctx, this, ()| {
+            Ok(this.0.borrow().ident.to_string())
```

Comment:
This will reallocate both in rust and then in lua, I think we can avoid it on the rust side by doing:
```suggestion
            Ok(this.0.borrow().ident.name.as_str().get())
```

Reply:
Good point, fixed now.